test/functional/lua/with_spec.lua

```diff
@@ -0,0 +1,296 @@
+      _G.win_lines = api.nvim_get_current_win()
+      vim.cmd "new"
+      _G.win_empty = api.nvim_get_current_win()
+      api.nvim_set_current_win(win_lines)
+      api.nvim_buf_set_lines(0, 0, -1, true, {"a", "b", "c"})
+    ]]
+
+    -- Start Visual in current window, redraw in other window with fewer lines.
+    -- Should be fixed by vim-patch:8.2.4018.
+    exec_lua [[
+      api.nvim_feedkeys(t "G<C-V>", "txn", false)
+      vim._with({win = win_empty}, function() vim.cmd "redraw" end)
+    ]]
+
+    -- Start Visual in current window, extend it in other window with more lines.
+    exec_lua [[
+      api.nvim_feedkeys(t "<Esc>gg", "txn", false)
+      api.nvim_set_current_win(win_empty)
+      api.nvim_feedkeys(t "gg<C-V>", "txn", false)
+      vim._with({win = win_lines}, function() api.nvim_feedkeys(t "G<C-V>", "txn", false) end)
+      vim.cmd "redraw"
+    ]]
+  end)
+
+  it('updates ruler if cursor moved', function()
+    local screen = Screen.new(30, 5)
+    screen:set_default_attr_ids {
+      [1] = { reverse = true },
+      [2] = { bold = true, reverse = true },
+    }
+    screen:attach()
+    exec_lua [[
+      _G.api = vim.api
+      vim.opt.ruler = true
+      local lines = {}
+      for i = 0, 499 do lines[#lines + 1] = tostring(i) end
+      api.nvim_buf_set_lines(0, 0, -1, true, lines)
+      api.nvim_win_set_cursor(0, {20, 0})
+      vim.cmd "split"
+      _G.win = api.nvim_get_current_win()
+      vim.cmd "wincmd w | redraw"
+    ]]
+    screen:expect [[
+      19                            |
+      {1:[No Name] [+]  20,1         3%}|
+      ^19                            |
+      {2:[No Name] [+]  20,1         3%}|
+                                    |
+    ]]
+    exec_lua [[
+      vim._with({win = win}, function() api.nvim_win_set_cursor(0, {100, 0}) end)
+      vim.cmd "redraw"
+    ]]
+    screen:expect [[
+      99                            |
+      {1:[No Name] [+]  100,1       19%}|
+      ^19                            |
+      {2:[No Name] [+]  20,1         3%}|
+                                    |
+    ]]
+  end)
+
+  it('can return values by reference', function()
+    eq(
+      { 7, 10 },
+      exec_lua [[
+      local val = {4, 10}
+      local ref = vim._with({win = 0}, function() return val end)
+      ref[1] = 7
+      return val
+    ]]
+    )
+  end)
+
+  it('layout in current tabpage does not affect windows in others', function()
+    command('tab split')
+    local t2_move_win = api.nvim_get_current_win()
+    command('vsplit')
+    local t2_other_win = api.nvim_get_current_win()
+    command('tabprevious')
+    matches('E36: Not enough room$', pcall_err(command, 'execute "split|"->repeat(&lines)'))
+    command('vsplit')
+
+    exec_lua('vim._with({win = ...}, function() vim.cmd.wincmd "J" end)', t2_move_win)
+    eq({ 'col', { { 'leaf', t2_other_win }, { 'leaf', t2_move_win } } }, fn.winlayout(2))
+  end)
+end)
+
+describe('vim._with {buf = }', function()
```

Comment:
I'd add tests that it doesn't trigger `{Buf,BufWin}{Enter,Leave}` events. Or at least be explicit and conscious about them (but better the former).